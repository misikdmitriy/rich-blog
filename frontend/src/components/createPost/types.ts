export interface CreatePostValues {
	title: string,
	shortUrl: string,
	description: string,
}
