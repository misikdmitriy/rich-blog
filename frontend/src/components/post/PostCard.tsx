import React from 'react';
import {
	Grid,
	Card,
	CardActionArea,
	CardContent,
	CardMedia,
	Typography,
} from '@mui/material';
import { Post } from '../../types/post';

interface PostCardProps {
	post: Post;
}

const PostCard = (props: PostCardProps) => {
	const { post } = props;

	return (
		<Grid item container>
			<Grid item xs={0} sm={2} />
			<Grid item xs={12} sm={8}>
				<CardActionArea component="a" href="#">
					<Card sx={{ display: 'flex' }}>
						<CardContent sx={{ flex: 1 }}>
							<Typography component="h2" variant="h5">
								{post.title}
							</Typography>
							<Typography variant="subtitle1" color="text.secondary">
								{post.date}
							</Typography>
							<Typography variant="subtitle1" paragraph>
								{post.description}
							</Typography>
							<Typography variant="subtitle1" color="primary">
								Continue reading...
							</Typography>
						</CardContent>
						<CardMedia
							component="img"
							sx={{ width: 260, display: { xs: 'none', sm: 'block' } }}
							image={post.image}
							alt={post.imageLabel}
						/>
					</Card>
				</CardActionArea>
			</Grid>
			<Grid item xs={0} sm={2} />
		</Grid>
	);
};

export default PostCard;
