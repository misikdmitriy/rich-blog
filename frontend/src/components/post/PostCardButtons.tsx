import React from 'react';
import {
	Box,
	IconButton,
} from '@mui/material';
import EditIcon from '@mui/icons-material/Edit';
import KeyIcon from '@mui/icons-material/Key';
import DeleteIcon from '@mui/icons-material/Delete';
import { Link } from 'react-router-dom';
import Dialog, { DialogConsumer } from '../dialog/Dialog';
import UnlockPostModal from '../unlockPost/UnlockPostModal';
import DeletePostModal from '../deletePost/DeletePostModal';
import { Post } from '../../types/post';

interface PostCardButtonsProps {
	post: Post;
	onPostDeleted?: () => void
	onPostUnlocked?: () => void
}

const PostCardButtons = (props: PostCardButtonsProps) => {
	const { post, onPostDeleted, onPostUnlocked } = props;

	return (
		<Box display="flex" alignItems="center" sx={{ pl: 1, pb: 1 }}>
			{!post.availableFor.includes('user') && (
				<Dialog>
					<DialogConsumer>
						{({ open }) => (
							<IconButton
								color="success"
								aria-label="unlock"
								onClick={(event) => {
									event.stopPropagation();
									open();
								}}
							>
								<KeyIcon />
							</IconButton>
						)}
					</DialogConsumer>

					<UnlockPostModal post={post} onPostUnlocked={onPostUnlocked} />
				</Dialog>
			)}

			<Link to={`/posts/edit/${post.id}`} style={{ textDecoration: 'none', color: 'inherit' }}>
				<IconButton
					color="warning"
					aria-label="edit"
				>
					<EditIcon />
				</IconButton>
			</Link>

			<Dialog>
				<DialogConsumer>
					{({ open }) => (
						<IconButton
							color="error"
							aria-label="delete"
							onClick={(event) => {
								event.stopPropagation();
								open();
							}}
						>
							<DeleteIcon />
						</IconButton>
					)}
				</DialogConsumer>

				<DeletePostModal post={post} onPostDeleted={onPostDeleted} />
			</Dialog>

		</Box>
	);
};

PostCardButtons.defaultProps = {
	onPostDeleted: () => {},
	onPostUnlocked: () => {},
};

export default PostCardButtons;
