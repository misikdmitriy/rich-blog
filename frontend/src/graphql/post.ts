import { gql } from '@apollo/client';

export const PostQuery = gql`
    query ($shortUrl: String!) {
        posts (filter: { shortUrl: $shortUrl }) {
            posts {
                id
                shortUrl
                body
                title
                description
                createdDate
            }
            hasNext
        }
    }`;
