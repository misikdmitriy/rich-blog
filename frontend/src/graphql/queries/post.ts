import { gql } from '@apollo/client';

export const PostByShortUrlQuery = gql`
    query ($shortUrl: String!) {
        posts (filter: { shortUrl: $shortUrl }) {
            posts {
                id
                shortUrl
                content
                title
                description
                createdDate
            }
        }
    }`;

export const PostByIdQuery = gql`
    query ($id: ID!) {
        posts (filter: { id: $id }) {
            posts {
                id
                title
                shortUrl
                description
                content
                image
                imageLabel
            }
        }
    }`;
