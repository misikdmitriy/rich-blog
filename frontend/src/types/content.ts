import { Variant } from '@mui/material/styles/createTypography';
import { NonEmptyArray } from './common';

export type ContentBase = {
    type: string
}

export type ContainerContent = ContentBase & Record<string, unknown> & {
    type: 'container',
    content: NonEmptyArray<ContentBase>
}

export type TextContent = ContentBase & {
    type: 'text',
    text: string
}

export type TypographyContent = ContentBase & Record<string, unknown> & {
    type: 'typography',
    variant: Variant,
    content: NonEmptyArray<ContentBase>
};
