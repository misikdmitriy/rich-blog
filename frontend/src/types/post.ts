export interface Post {
	id: string,
	shortUrl: string,
	title: string
	description: string
	image: string
	imageLabel: string
	body: string
	createdDate: Date
}

export interface PostsResponse {
	posts: Post[]
	hasNext: boolean
}
