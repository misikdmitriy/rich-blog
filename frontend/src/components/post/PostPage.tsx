import React from 'react';
import { useQuery } from '@apollo/client';
import {
	CircularProgress, Typography,
} from '@mui/material';
import { useParams } from 'react-router-dom';
import { PostQuery } from '../../graphql/post';
import { Post } from '../../types/post';
import NotFound from '../notfound/NotFound';

interface OptionalPostResult {
    posts: [Post?]
}

const PostPage = () => {
	const { shortUrl } = useParams();
	const {
		data: {
			posts: {
				posts: [post] = [],
			} = {},
		} = {}, loading,
	} = useQuery<{ posts: OptionalPostResult }>(PostQuery, {
		variables: {
			shortUrl,
		},
	});

	if (!loading && !post) {
		return <NotFound />;
	}

	if (loading) {
		return <CircularProgress />;
	}

	return (
		<div style={{ margin: '1em 4em' }}>
			<div>
				<Typography variant="h2" component="h1">
					{post?.title}
				</Typography>
			</div>
			<div dangerouslySetInnerHTML={{ __html: post?.body || '' }} />
		</div>
	);
};

export default PostPage;
