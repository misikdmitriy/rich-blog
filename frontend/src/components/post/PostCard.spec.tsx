import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { Post } from '../../types/post';
import PostCard from './PostCard';

describe('PostCard', () => {
	const post: Post = {
		id: '62d04cc36d3881c902280116',
		shortUrl: '/short-url',
		title: 'Random Title',
		description: 'Some description',
		image: 'http://image.com/jpg',
		imageLabel: 'image alt',
		content: '{}',
		createdDate: new Date(Date.parse('Dec 25, 1995')),
		availableFor: ['user'],
	};

	it('renders without crashing', () => {
		const tree = renderer
			.create(<MemoryRouter><PostCard post={post} /></MemoryRouter>)
			.toJSON();

		expect(tree).toMatchSnapshot();
	});
});
