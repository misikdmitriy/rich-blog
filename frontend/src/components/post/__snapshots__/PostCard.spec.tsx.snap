// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostCard renders without crashing 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-ysuk8h-MuiPaper-root-MuiCard-root"
>
  <a
    href="/posts/id//short-url"
    onClick={[Function]}
    style={
      Object {
        "color": "inherit",
        "textDecoration": "none",
      }
    }
  >
    <button
      className="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
      disabled={false}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <img
        alt="image alt"
        className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-q0pqgy-MuiCardMedia-root"
        src="http://image.com/jpg"
      />
      <div
        className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <h3
          className="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
        >
          Random Title
        </h3>
        <p
          className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
        >
          December 25th 1995
        </p>
        <p
          className="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph css-z2eky3-MuiTypography-root"
        >
          Some description
        </p>
        <p
          className="MuiTypography-root MuiTypography-body1 css-vkc4v8-MuiTypography-root"
        >
          Continue reading...
        </p>
      </div>
      <span
        className="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
      />
    </button>
  </a>
</div>
`;
